contracts/payment/SponsorPayment.sol

```diff
@@ -0,0 +1,519 @@
+
+    // Event emitted when a deposit is made ERC20
+    event ERC20Deposit(address indexed sponsor, address indexed token, uint256 amount);
+
+    // Event emitted when a deposit is made
+    event Deposit(address indexed sponsor, uint256 amount);
+
+    // Event emitted when a withdrawal is made
+    event Withdrawal(address indexed sponsor, uint256 amount);
+
+    // Event emitted when a withdrawal is made
+    event ERC20Withdrawal(address indexed sponsor, address indexed token, uint256 amount);
+
+    // Event emitted when a withdrawal request is made
+    event WithdrawalRequested(address indexed sponsor, uint256 amount, uint256 lockTime);
+
+    // Event emitted when a withdrawal request is made ERC20
+    event ERC20WithdrawalRequested(
+        address indexed sponsor,
+        address indexed token,
+        uint256 amount,
+        uint256 lockTime
+    );
+
+    // Event emitted when a withdrawal request is cancelled
+    event WithdrawalCancelled(address indexed sponsor, uint256 amount);
+
+    // Event emitted when a withdrawal request is cancelled ERC20
+    event ERC20WithdrawalCancelled(address indexed sponsor, address indexed token, uint256 amount);
+
+    // Event emitted when a payment is claimed
+    event PaymentClaimed(address indexed recipient, uint256 indexed nonce, uint256 amount);
+
+    // Event emitted when a payment is claimed
+    event ERC20PaymentClaimed(
+        address indexed recipient,
+        uint256 indexed nonce,
+        address indexed token,
+        uint256 amount
+    );
+
+    // Event emitted when the owner withdraws their balance
+    event OwnerBalanceWithdrawn(uint256 amount);
+
+    // Event emitted when the owner withdraws their balance ERC20
+    event ERC20OwnerBalanceWithdrawn(uint256 amount);
+
+    /**
+     * @dev Payment details used in claim logic.
+     */
+    struct ERC20SponsorPaymentInfo {
+        address recipient;
+        uint256 amount;
+        address token;
+        uint256 nonce;
+        uint256 expiration;
+        bytes metadata;
+    }
+
+    /**
+     * @dev Payment details used in claim logic.
+     */
+    struct SponsorPaymentInfo {
+        address recipient;
+        uint256 amount;
+        uint256 nonce;
+        uint256 expiration;
+        bytes metadata;
+    }
+
+    struct WithdrawalRequest {
+        uint256 amount;
+        uint256 lockTime;
+        bool exists;
+    }
+
+    /**
+     * @dev Main storage $tructure for the contract
+     */
+    struct SponsorPaymentStorage {
+        mapping(address => mapping(address => uint256)) balances; // sponsor => token => balance
+        mapping(address => mapping(address => WithdrawalRequest)) withdrawalRequests; // sponsor => token => request
+        mapping(bytes32 => bool) isWithdrawn;
+        uint8 ownerPercentage;
+        uint256 withdrawalDelay;
+    }
+
+    string public constant VERSION = "1.0.0";
+
+    bytes32 public constant ERC20_PAYMENT_CLAIM_DATA_TYPE_HASH =
```

Comment:
What is a meaning of the word 'claim'? do we talk about payment for credentials? or iden3 core claims? or claims as attributes of credential?

I would remove this word at all. It is clear that this is contract is for payments for smth on the general level

if it is about 'please give me money' from deposit we can use word' obtain' payment to avoid confusion.

it is for discussion.

Reply:
if it is about 'please give me money'
I don't think that 'obtain' payment suits here also

Reply:
let's call it as struct we hash - ERC_20_SPONSO_PAYMENT_INFO_TYPE_HASH

Reply:
renamed